README.md
[![Build Status](https://dev.azure.com/devrel/chris-testing/_apis/build/status/softchris.yaml-server?branchName=master)](https://dev.azure.com/devrel/chris-testing/_build/latest?definitionId=23&branchName=master)
![Coverage](./badges/coverage.svg)
[![npm version](https://badge.fury.io/js/yaml-server.svg)](https://www.npmjs.com/package/yaml-server)
[![npm downloads](https://img.shields.io/npm/dm/yaml-server?color=blue&label=npm%20downloads&style=flat-square)](https://www.npmjs.com/package/yaml-server)
[![The MIT License](https://img.shields.io/badge/license-MIT-orange.svg?color=blue&style=flat-square)](http://opensource.org/licenses/MIT)

## Table of Contents

- [About](#about)
- [Features](#features)
- [Install](#install)
- [Run](#run)
- [Routes](#routes)

## About

Recognition, this project wouldn't be here with out the great `json-server`. I thought to myself that JSON was a little verbose. So I created `yaml-server` so you can have a Mock REST API based on a YAML file instead.

`yaml-server` is a command line tool that create a REST server based on a YAML file.

![Application running](yaml-server.png)

## Features

- **RESTful API** Do HTTP requests towards a Mock API using GET, PUT, POST and DELETE created off of a `db.yml` file.
- **Filter** your GET calls with query parameters `page` and `pageSize`, example:

    ```bash
    /products?page=1&pageSize=10
    ```

- **JSON support**, yes you can have your database in JSON as well. All you need is to specify the `--dialect` argument like so:

   ```bash
   npx yaml-server --dialect=json
   ```

   The above will look after a `db.json` file at the root. You override where it looks for this if you specify `--database` like for example:

   ```bash
   npx yaml-server --dialect=json --database ./db/db.json
   ```

   Above you need to ensure the `db.json` is located in sub directory `db` as seen from the root.

- **Create new resource**, make a POST call with the following format `/<new resource>/new`, example:

    ```bash
    /kittens/new
    ```

    Ensure you have a payload as well, looking like for example `{ title: 'paw paw' }`

- **Sorting, by order and key**, you can sort a resource response using query parameters `sortOrder` and `sortKey`. Assume we have the resource `/kittens` where one kitten object looks like so `{ id: 1, name: 'paws' }` and the entire resource looks like so:

    ```javascript
    [{
      id: 1,
      name: 'paws'
    }, {
      id: 2,
      name: 'alpha paw'
    }]
    ```

    Use sorting by appending `sortOrder` and `sortKey` like below:

    ```bash
    /kittens?sortOrder=ASC&sortKey=name
    ```

    This would give the response:

    ```javascript
    [{
      id: 2,
      name: 'alpha paw'
    }, {
      id: 1,
      name: 'paws'
    }]
    ```

- **browser autostart**, the Browser auto starts at `http://locallhost:<selected port>/info`. Should you not wish to have that behavior, you can shut it off like so:

   ```bash
   npx yaml-server --autoStart=off
   ```

- **Static file server**

  By default a static file server is starting up to host your files at root directory. You can change that by specifying `--static`. Here's how you would do that:

   ```bash
   npx yaml-server --static=public
   ```

   The above would start a static file server from the sub folder `public`.

- **Hot reload**

  The server will restart if you make changes to your database file. No need for closing and starting the server after a database change. Should you not wish that behavior, you can shut it off with:

  ```bash
  npx yaml-server --hotReload=off
  ```

## Install

Either install it globally with:

```bash
npm install -g yaml-server
```

OR use `NPX`

```bash
npx yaml-server --port 3000 --database ./db.yml
```

## Run

1. Create a `db.yml`.
1. Give `db.yml` an example content, for example:

   ```yaml
   products:
     - id: 1
       name: tomato
     - id: 2
       name: lettuce
   orders:
    - id: 1
      name: order1
    - id: 2
      name: order2
   ```

1. There are two ways to start:
   1. **Quick start**, run `npx yaml-server`, this will start a server on `http://localhost:3000` and base it off a `db.yml` at the project root that you created.
   1. **With parameters**, You can also configure like so `npx yaml-server --port 8000 --database ./db/mydb.yml` (If you place db file under `./db/mydb.yml`)

### See your routes

Open up a browser and navigate to `http://localhost:<your port>/info`. Default port is `3000`, if you specified port use that as port instead.

The page at route `http://localhost:<port>/info` will tell you what routes and operations are available. Here's a typical response for the default page:

```output
Welcome to YAML Server

Routes available are:

 GET /products
 GET /products/:id
 PUT /products
 DELETE /products/:id

 GET /orders
 GET /orders/:id
 PUT /orders
 DELETE /orders/:id
```

## Routes

Routes are created from a YAML file. The default value is  `db.yml`. You can name it whatever you want though.

Routes are first level elements. Consider the following example file:

```yml
# db.yml
products:
  - id: 1
    name: tomato
  - id: 2
    name: lettuce
orders:
  - id: 1
    name: order1
  - id: 2
    name: order2
```

This will produce routes `/products`, `/orders`. Below is a table of supported operations with `products` as example resource. The same operations are also supports for `orders/`.

| VERB     |Route          | Input      | Output             |
|----------|---------------|------------|--------------------|
| GET      | /products     | *None*     | **Array**          |
| GET      | /products/:id |  **e.g 3** | **Object**         |
| POST     | /products     | **object** | **Created object** |
| PUT      | /products     | **object** | **Updated object** |
| DELETE   | /products/:id | **e.g 3**  | **Deleted object** |


CHANGELOG.md
# Changelog

## 1.2.0 - 2020-05-20

- Adding support for query parameters `page` and `pageSize`

## 1.2.1 - 2020-05-20

- Made run instructions clearer

## 1.3.1 - 2020-05-20

- Adding support for new resource creation `/<resource>/new`

## 1.3.2 - 2020-05-20

- Test added to ensure no new resource is created with `/<resource>/new` if resource already exist

## 1.3.3 - 2020-05-21

- Adding CORS, all requests are CORS enabled

## 1.4.4 - 2020-05-21

- Adding ESLint

## 1.5.0 - 2020-05-21

- Adding CI with Azure Devops

## 1.6.0 - 2020-05-21

- Added ability to sort.

## 1.7.0 - 2020-05-21

- Static file hosting, use `--static` to specify what directory, otherwise root is used.

## 1.8.0 - 2020-05-21

- Auto start browser at `http://localhost:<selected port>/info`.

## 1.9.0 - 2020-05-29

- Hot reload, you can now edit the db file and the server will restart by itself.


src/__tests__/server.js
const { createServer, getHttpServer } = require("../server");
const path = require("path");
const fs = require("fs");
const supertest = require("supertest");
const jsonOption = {
  dialect: 'json',
  dbFilename: 'db.json'
};

const ymlOption = {
  dialect: 'yaml',
  dbFilename: 'db.yml'
};

function createOption(option) {
  if (option === 'yaml') {
    return ymlOption
  } else if (option === 'json') {
    return jsonOption;
  } else {
    throw new Error(`Unknown option ${option}`)
  }
}

const selectedOption = createOption('json');

const server = createServer(
  3000, 
  path.join(__dirname, selectedOption.dbFilename), 
  ".", 
  false, 
  selectedOption.dialect
);

const request = supertest(server);

function restore(dialect) {
  if (dialect === 'yaml') {
    const original = fs.readFileSync(path.join(__dirname, "original-db.yml"));
    fs.writeFileSync(path.join(__dirname, "db.yml"), original);
  } else if (dialect === 'json') {
    const original = fs.readFileSync(path.join(__dirname, "original-db.json"));
    fs.writeFileSync(path.join(__dirname, "db.json"), original);
  }
}

describe("server", () => {
  afterAll(async(done) => {
    // restore yml
    // TODO, ensure we have an original based on dialect
    // const original = fs.readFileSync(path.join(__dirname, "original-db.yml"));
    // fs.writeFileSync(path.join(__dirname, "db.yml"), original);
    restore(selectedOption.dialect);
    getHttpServer().close(() => {
      console.log("server closed!");
      done();
    });
  });

  test("should return products", async(done) => {
    const products = [{
      id: 1,
      name: "tomato"
    }, {
      id: 2,
      name: "lettuce"
    }];

    const res = await request.get("/products");
    expect(res.status).toBe(200);
    expect(res.body).toEqual(products);
    done();
  });

  test("should sort the data based on sortOrder and sortKey", async () => {
    const expected = [{ id: 2, name: 'lettuce' }, { id: 1, name: 'tomato' }];

    const res = await request.get('/products?sortOrder=ASC&sortKey=name')
    expect(expected).toEqual(res.body);
  })

  test("should sort the data based on sortOrder and sortKey - DESCENDING", async () => {
    const expected = [{ id: 1, name: 'tomato' }, { id: 2, name: 'lettuce' }];

    const res = await request.get('/products?sortOrder=DESC&sortKey=name')
    expect(expected).toEqual(res.body);
  })

  test("should sort with ascending when sortOrder is missing but sortKey is present", async () => {
    const expected = [{ id: 2, name: 'lettuce' }, { id: 1, name: 'tomato' }];

    const res = await request.get('/products?sortKey=name')
    expect(expected).toEqual(res.body);
  })

  test("should sort not respect sortOrder when sortOrder value is NOT ASC or DESC", async () => {
    const expected = [{ id: 1, name: 'tomato' }, { id: 2, name: 'lettuce' }];

    const res = await request.get('/products?sortOrder=abc&sortKey=name')
    expect(expected).toEqual(res.body);
  })

  test("should respond with 400 when sortKey is not a valid column", async () => {
    const expected = [{ id: 1, name: 'tomato' }, { id: 2, name: 'lettuce' }];

    const res = await request.get('/products?sortOrder=abc&sortKey=notValidKey')

    expect(res.status).toBe(400)
    expect(res.text).toBe('notValidKey is not a valid sort key')
  })

  test("should filter by query parameters", async() => {
    const firstItem = { id: 1, name: "tomato" };
    const secondItem = { id: 2, name: "lettuce" };

    let res = await request.get("/products?page=1&pageSize=1");
    let [ item ] = res.body;
    expect(res.body).toHaveLength(1);
    expect(item).toEqual(firstItem);

    res = await request.get("/products/?page=2&pageSize=1");
    item = res.body[0];
    expect(res.body).toHaveLength(1);
    expect(item).toEqual(secondItem);
  });

  test("should NOT respect filter when query param missing", async() => {
    const res = await request.get("/products?page=1");
    expect(res.body).toHaveLength(2);
  });

  test("should NOT respect filter when query param has wrong value type", async () => {
    const res = await request.get("/products?page=1&pageSize=abc");
    expect(res.body).toHaveLength(2);
  });

  test("should return intro text on default route", async() => {
    const res = await request.get("/info");
    expect(res.text).toMatch(/Welcome to YAML Server/);
  });

  test("should return a product", async(done) => {
    const product = { id: 1, name: "tomato" };
    const res = await request.get("/products/1");
    expect(res.status).toBe(200);
    expect(res.body).toEqual(product);
    done();
  });

  test("should return 404 resource not found", async (done) => {
    const res = await request.get("/products/3");
    expect(res.status).toBe(404);
    done();
  });

  test("should add product to /products", async(done) => {
    const createdRecord = { id: 3, name: "cucumber" };

    const res = await request
      .post("/products")
      .send({ name : "cucumber" });
    expect(res.status).toBe(201);
    expect(res.body).toEqual(createdRecord);
    done();
  });

  test("should update product", async(done) => {
    const changeTo = { id : 3, name: "gurkin" };
    let res = await request
      .put("/products")
      .send({ id: 3, name: "gurkin" });
    expect(res.status).toBe(200);
    expect(res.body).toEqual(changeTo);

    res = await request.get("/products/3");
    expect(changeTo).toEqual(res.body);
    done();
  });

  test("should return 404 and error message when trying to update non existing item", async() => {
    const nonExistingItem = { id: 99, name: "unknown" };
    let res = await request
      .put("/products")
      .send(nonExistingItem);
    expect(res.status).toBe(404);
    expect(res.text).toBe("Item not found with ID" + nonExistingItem.id);
  });

  test("should delete product", async(done) => {
    const deletedItem = { id: 3, name: "gurkin" };
    let res = await request.delete("/products/3");
    expect(res.status).toBe(200);
    expect(res.body).toEqual(deletedItem);

    res = await request.get("/products/3");
    expect(res.status).toBe(404);
    done();
  });

  test("should respect existing schema on POST", async() => {
    let res = await request
      .post("/products")
      .send({ title: "should not work" });
    
    expect(res.status).toBe(400);
  });

  test("should create a new resource based on a /<resource>/new/ call", async() => {
    const kitten = { id: 1, title: "paw paw" };
    let res = await request
      .post("/kittens/new")
      .send({ title: "paw paw" });

    expect(res.status).toBe(201);  
    expect(res.body).toEqual(kitten);

    res = await request.get("/kittens");
    expect(res.status).toBe(200);
    expect(res.body).toEqual(kitten);
  });

  test("should NOT create a new resource /<resource>/new/ call when resource already exist", async() => {
    let res = await request
      .post("/kittens/new")
      .send({ title: "paw paw" });

    expect(res.status).toBe(400);
    expect(res.text).toBe("/kittens already exist");
  });
});

src/index.js
#!/usr/bin/env node
const argv = require("yargs").argv;
const fs = require('fs');

function getDefaultByDialect(dialect) {
  switch (dialect) {
    case 'yaml':
      return 'db.yml'
    case 'json':
      return 'db.json'
    default:
      return '';
  }
}

const { createServer, getHttpServer } = require("./server");
const dialect = argv.dialect || 'yaml';
const { dbPath } = require("./config")(getDefaultByDialect(dialect));

const port = argv.port || 3000;
const database = argv.database || dbPath;
const autoStart = argv.autoStart === 'off' ? false : true;
const hotReload = argv.hotReload === "off" ? false : true;

createServer(port, database, argv.static, autoStart, dialect);

if (hotReload) {
  fs.watchFile(database, (curr, prev) => {
    console.log("Database changed");
    getHttpServer().close(() => {
      console.log("Restarting the server");
      createServer(port, database, argv.static, autoStart);
    });
  });
}
  



src/server.js
const YAML = require("yaml");
const fs = require("fs");
var bodyParser = require("body-parser");
const cors = require("cors");
const chalk = require("chalk");
const express = require("express");
const { toJson, transform } = require('./fileHelper');
const opn = require('opn');

let httpServer;

function createServer(portNumber, dbPath, staticDirectory, autoStart = true, dialect = 'yaml') {  
  let app = express();
  app.use(cors());
  
  const port = portNumber || 3000;

  app.use(bodyParser.json());
  
  let routes = [];
  let json;
  if (!fs.existsSync(dbPath)) {
    console.log(chalk.redBright(`DB does not exist at path "${dbPath}"`));
  } else {
    const content = fs.readFileSync(dbPath, { encoding: "utf8" });
    json = toJson(content, dialect)
    routes = Object.keys(json);
  }
  
  let staticPath = staticDirectory || ".";
  app.use(express.static(staticPath));
  console.log(chalk.blueBright(`Static files served from ${staticPath}/`));

  const routesString = routes.reduce((acc, curr) => {
    return `${acc} GET /${curr} \n GET /${curr}/:id \n PUT /${curr} \n DELETE /${curr}/:id \n\n`;
  }, "");

  const routeDefault = (req, res) => res.send(`Welcome to YAML Server \n
Routes available are: \n
${routesString}
`);

  function setupRoutesForResource(route) {
    app.get(`/${route}`, (req, res) => routeGet(req, res, route));
    app.get(`/${route}/:id`, (req, res) => routeGetWithParam(req, res, route));
    app.post("/:newRoute/new", (req, res) => routeNewResource(req, res, route));
    app.post(`/${route}`, (req, res) => routePost(req, res, route));
    app.put(`/${route}`, (req, res) => routePut(req, res, route));
    app.delete(`/${route}/:id`, (req, res) => routeDelete(req, res, route));
  } 

  function routeGet(req, res, route) {
    const page = req.query.page;
    const pageSize = req.query.pageSize;
    const sortOrder = req.query.sortOrder;
    const sortKey = req.query.sortKey;
    const sortOrders = [ 'ASC', 'DESC' ];

    const sortAscending = (a, b) => {
      console.log('sort ascending')
      if (a[sortKey] > b[sortKey]) {
        return 1;
      } else if (a[sortKey] < b[sortKey]) {
        return -1;
      }
      return 0;
    }

    const sortDescending = (a, b) => {
      if (a[sortKey] > b[sortKey]) {
        return -1;
      } else if (a[sortKey] < b[sortKey]) {
        return 1;
      }
      return 0;
    }

    if (sortKey && json[route][0][sortKey] === undefined) {
      res.statusCode = 400;
      res.send(`${sortKey} is not a valid sort key`);
    }

    if (sortOrder && sortOrders.includes(sortOrder)) {
      const sortMethod = sortOrder === 'ASC'? sortAscending : sortDescending;
      const copyArr = [ ...json[route]];
      copyArr.sort(sortMethod)
      res.json(copyArr)
    } else if (!sortOrder && sortKey) {
      const copyArr = [...json[route]];
      copyArr.sort(sortAscending)
      res.json(copyArr)
    }

    if (/\d+/.test(page) && /\d+/.test(pageSize)) {
      const pageNo = +page;
      const pageSizeNo = +pageSize;
      const start = pageSizeNo * (+pageNo - 1);
      const end = start + pageSizeNo;
      res.json(json[route].slice(start, end));
    }
    res.json(json[route]);
  }

  function routeGetWithParam(req, res, route) {
    const foundItem = json[route].find((item) => item.id == req.params.id);
    if (!foundItem) {
      res.statusCode = 404;
      res.json({});
    } else {
      res.json(foundItem);
    }
  }

  function routeNewResource(req, res) {
    const { newRoute } = req.params;

    if (!json[newRoute]) {
      json[newRoute] = { ...req.body, id: 1 };
      setupRoutesForResource(newRoute);
      fs.writeFileSync(dbPath, transform(json, dialect));
      res.statusCode = 201;
      res.json({ ...req.body, id: 1 });
    } else {
      res.statusCode = 400;
      res.send(`/${newRoute} already exist`);
    }
  }

  function routePost(req, res, route) {
    const posted = { ...req.body, id: 0 };

    if (json[route].length > 0) {
      const [firstItem] = json[route];
      const props = [...Object.keys(firstItem)].sort();
      const postedProps = Object.keys(posted).sort();

      if (JSON.stringify(props) !== JSON.stringify(postedProps)) {
        res.statusCode = 400;
        res.send("");
      }
    }

    const insertObject = { ...posted, id: json[route].length + 1 };

    json[route].push(insertObject);

    fs.writeFileSync(dbPath, transform(json, dialect));
    res.statusCode = 201;
    res.json(insertObject);
  }

  function routePut(req, res, route) {
    const posted = req.body;
    let foundItem;
    json[route] = json[route].map((item) => {
      if (item.id === +posted.id) {
        foundItem = item;
        return { ...item, ...posted };
      }
      return item;
    });
    if (foundItem) {
      fs.writeFileSync(dbPath, transform(json, dialect));
      res.json({ ...foundItem, ...posted });
    } else {
      res.statusCode = 404;
      res.send("Item not found with ID" + posted.id);
    }
  }

  function routeDelete(req, res, route) {
      let deletedItem;
      json[route] = json[route].filter((item) => {
        if (item.id === +req.params.id) {
          deletedItem = item;
        }
        return item.id !== +req.params.id;
      });
      fs.writeFileSync(dbPath, transform(json, dialect));
      return res.json(deletedItem);
    }

  app.get("/info", routeDefault);


  routes.forEach(setupRoutesForResource);

  httpServer = app.listen(port, async() => {
    console.log(`Example app listening on port ${port}!`);
    console.log(chalk.greenBright(routesString));
    if (autoStart) {
      const ref = await opn(`http://localhost:${port}/info`);
    }
  });
  return app;
}

function getHttpServer() {
  return httpServer;
}

module.exports = {
  createServer,
  getHttpServer
};
